### DISCUSSION TITLE
How to update ui::RichText’s string content after creation?

### DISCUSSION THREAD

**Discussion by antn9x** [axmolengine/axmol#1166] [General]
I’d like to change my RichText’s content from "abc", to "def", but it’s not clear how. There’s no setString equivalent, and the closest is just calling initWithXML, but that comes with its own issues since the old renderer nodes are still present.
```
auto rich_text = cocos2d::ui::RichText::createWithXML("abc");
rich_text->setString("def"); //setString doesn't exist
rich_text->initWithXML("def"); //incorrectly renders both 'abc' and 'def' next to each other
```

**Comment by rh101** [comment]
It doesn't work that way.  `RichText` is a container for the different elements that are used to construct it, such as `ui::RichElementText`, which is the one that handles the text.  The only way is to re-create the entire `RichText`.

If you are not dealing with formatted text (with XML tags) or embedded graphics, then `ui::Label` is the better option, which does allow string updates.

EDIT: Sorry, I totally missed the point of the post. The specific issue is that it's not clearing the contents, but appending the new text to the existing contents; it wasn't about setting a string as `ui::Label` does.

**Comment by aismann** [comment]
As  @rh101 has written:
> The only way is to re-create the entire `RichText`.

You can try re-create with: (removeChild() and addChild())

```
rich_text = ax::ui::RichText::createWithXML("abc");
_widget->addChild(rich_text);
[...lot of code...]

_widget->removeChild(rich_text);
rich_text = ax::ui::RichText::createWithXML("def");
_widget->addChild(rich_text);
```

**Comment by antn9x** [comment]
I don't know why there was no implement for `setString` method, i think it posible and easy for update richtext string.

**Reply by rh101** [reply]
~~Once again, it doesn't work the way you may think it does.  The `RichText` can be made up of **many** text elements, so how would the `RichText` know which one it needs to update?  If you have a simple string, then use `ui::Label`, because that will allow you to update the string, since there is only one in there.~~

~~Have a look at the code for `RichText`, and it will make sense as to why that is not possible.~~

~~The only way to really fix this is to introduce both `initWithPlainText` and `setPlainText` methods which would only allow a single `RichElementText` element to be constructed (so no XML formatted text).  The way it would work is any time `setPlainText` is called, it would simply delete the entire element list, and then add a new `RichElementText` into it.  This can be made more efficient if it checks how many elements are inside the list, so if only one exists, and is of type `RichElementText`, then it would simply update it.  `RichElementText` would also need to be modified to allow for this, because it currently does not support this. Re-using current elements cannot be done, because newline characters are their own elements, so the entire element list will need to be cleared and recreated.~~

EDIT: ~~`initWithXML` may have a bug in it if it is not overwriting the current XML formatted `RichText`, so I'll look into that too, but the issue is that the current behaviour may be as expected, so that will be a separate PR to the other changes in case it's not desired behaviour.~~ @antn9x has a better option for fixing this by introducing the `removeAllElements()` method.

EDIT: I totally misunderstood the initial issue. It's just a matter of clearing the current contents and re-initializing with the new XML string.

**Reply by antn9x** [reply]
Of course I use `Rich Text` because of XML formatted text.

**Reply by rh101** [reply]
Then the main issue is that `initWithXML` is not over-writing the current contents, correct?

**Reply by aismann** [reply]
> Then the main issue is that `initWithXML` is not over-writing the current contents, correct?

Another question is: What is the reason for method `initWithXML` ?
I belive "update/replace' isn't it

**Reply by rh101** [reply]
> > Then the main issue is that `initWithXML` is not over-writing the current contents, correct?
> 
> Another question is: What is the reason for method `initWithXML` ? I belive "update/replace' isn't it

Well, it's technically initializing the RichText, not updating it, so the method name is not incorrect.  Using "replace" wouldn't make sense, and "update" would only make sense if the current `RichText` contents are cleared (which is not the case at the moment).

**Comment by antn9x** [comment]
I created  [PR](https://github.com/axmolengine/axmol/pull/1167) for this isssue. Can you help to merge?
Then I can do like this:
```
    richText->removeAllElements();
    richText->initWithXML(myNewXML);
```

**Reply by rh101** [reply]
That's a good change, since it does not affect current behaviour of `RichText`, and gives the developer the option to clear it before calling `initWithXML`.

**Comment by aismann** [comment]
Is this not the better/same way?
```
rich_text = ax::ui::RichText::createWithXML("abc");
_widget->addChild(rich_text);
[...lot of code...]

_widget->removeChild(rich_text);
rich_text = ax::ui::RichText::createWithXML("def");
_widget->addChild(rich_text);
```

and `initWithXML `is the 'same' as  `createWithXML`

**Reply by rh101** [reply]
That would mean you're freeing the current memory associated with the `RichText`, and recreating another one, then re-adding it to the scene, which is more processing.  By using `removeAllElements`, it would mean no need for the extra overhead of removing and re-adding to the parent node, and less memory release/new calls.

> and `initWithXML `is the 'same' as `createWithXML`

Not sure what you mean. `createWithXML` creates a new instance of `RichText`, but `initWithXML` sets up the contents of the `RichText`, which are two very different functions.

**Reply by halx99** [reply]
Will implement a setString for RichText

**Reply by halx99** [reply]
I already test on my local machine:

<img width="665" alt="image" src="https://user-images.githubusercontent.com/6977319/230016108-2b3f21ca-4711-4663-9447-92a782ca2510.png">
<img width="843" alt="image" src="https://user-images.githubusercontent.com/6977319/230016243-2dab6f1e-462e-4e80-a834-b1886a3ee617.png">

The update code looks like:

![Uploading image.png…]()

**Comment by halx99** [comment]
Done with commit: https://github.com/axmolengine/axmol/commit/a4e75d47a77b0e6ccb7a402e839a25dcb033d4c2
